### ISSUE TITLE
SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch

### ISSUE THREAD

**Issue by itn3rd77** [danielperna84/pyhomematic#239]
Make sure you are running the latest version of pyhomematic (and Home Assistant) before reporting an issue.

_If you encounter the problem when using Home Assistant, please fill out the following information_

**Home Assistant version (if applicable):**
0.91.3

**Problem-relevant `configuration.yaml` entries and steps to reproduce:**
```yaml
interfaces:
  rf:
    host: homematic-raspi.local.home
    port: 42001
    ssl: true
    verify_ssl: true
    username: Guest
    password: ********
  ip:
    host: homematic-raspi.local.home
    port: 42010
    ssl: true
    verify_ssl: true
    username: Guest
    password: ********
hosts:
  raspberrymatic:
      host: homematic-raspi.local.home
      username: Guest
      password: ********
```

Use the following configuration for debug-output of the related components:
```yaml
logs:
  homeassistant.components.homematic: debug
  pyhomematic: debug
```

```
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic.connection] HMConnection: Creating server object
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] ServerThread.__init__
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] __init__: Creating proxies
2019-04-16 07:58:42 INFO (SyncWorker_19) [pyhomematic._hm] Creating proxy rf. Connecting to 192.168.100.74:42001
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Getting local ip
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Got local ip 192.168.100.2
2019-04-16 07:58:42 INFO (SyncWorker_19) [pyhomematic._hm] Creating proxy ip. Connecting to 192.168.100.74:42010
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Getting local ip
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Got local ip 192.168.100.2
2019-04-16 07:58:42 INFO (SyncWorker_19) [pyhomematic._hm] Creating proxy raspberrymatic. Connecting to 192.168.100.74:2001
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Getting local ip
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] LockingServerProxy.__init__: Got local ip 192.168.100.2
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] RPCFunctions.__init__
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] RPCFunctions.__init__: iterating proxy = homeassistant-rf
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] RPCFunctions.__init__: iterating proxy = homeassistant-ip
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] RPCFunctions.__init__: iterating proxy = homeassistant-raspberrymatic
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] ServerThread.__init__: Setting up server
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] ServerThread.__init__: Registering RPC functions
2019-04-16 07:58:42 INFO (Thread-3) [pyhomematic._hm] Starting server at http://0.0.0.0:47051
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] ServerThread.proxyInit: init('http://192.168.100.2:47051', 'homeassistant-rf')
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] proxyInit: Exception: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch, certificate is not valid for '192.168.100.74'. (_ssl.c:1056)
2019-04-16 07:58:42 WARNING (SyncWorker_19) [pyhomematic._hm] Failed to initialize proxy
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] ServerThread.proxyInit: init('http://192.168.100.2:47051', 'homeassistant-ip')
2019-04-16 07:58:42 DEBUG (SyncWorker_19) [pyhomematic._hm] proxyInit: Exception: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch, certificate is not valid for '192.168.100.74'. (_ssl.c:1056)
2019-04-16 07:58:42 WARNING (SyncWorker_19) [pyhomematic._hm] Failed to initialize proxy
```

**Describe the bug**
I run a RaspberryMatic with HTTPS enabled and wanted to set up Home Assistant homematic with ssl and verify_ssl enabled. I run my own PKI and the RaspberryMatic box has a server certificate signed by my PKI. I have exported the environment variable ```REQUESTS_CA_BUNDLE``` pointing to my CA certificate and a sample HTTPS request with Python Requests work fine. But pyhomematic is struggeling with ```Exception: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch, certificate is not valid for '192.168.100.74'```. The problem is pyhomematic  is using the ip-address 192.168.100.74 instead of the hostname ```homematic-raspi.local.home``` as configured. Putting the ip address in the certificate is an option but not a good choice I would like to go. I think when working with SSL it makes sense to work with hostnames instead of ip addresses. Or am I wrong?

**To Reproduce**

**Expected behavior**
Please switch pyhomematic  to use the configured value from the ```host:``` field instead of always using the ip address. 

**Screenshots**

**Additional context**

**Comment by itn3rd77** [comment]
**Update**

To verify my assumption I did a quick hack in [_hm.py#L515](https://github.com/danielperna84/pyhomematic/blob/8b91f3e84c83f05d289c740d507293a0d6759d8e/pyhomematic/_hm.py#L515) changing from ```socket.gethostbyname(host['ip'])``` to ```host['ip'] = socket.gethostbyname(host['ip'])```.

Using hostname with ssl and verify_ssl enabled work fine!

**Comment by danielperna84** [comment]
Actually I believe the problem is in Home Assistant. Have a look at [this](https://github.com/home-assistant/home-assistant/blob/177ae3fd322e116f460c3b23d7c0a7a74dc90cd2/homeassistant/components/homematic/__init__.py#L265). Any host (hostname or IP) get's converted to an IP. So I wonder how your change fixed the issue. Home Assistant always passes IP addresses. 🤔 
Back then when I added SSL support I only tested it within pyhomematic, and there it was working. So I think removing the `socket.gethostbyname` from the component is all that needs to be done. Could you test that please?

**Comment by itn3rd77** [comment]
Hi,

yes the root cause is in the Home Assistant component at lines [homeassistant/components/homematic/__init__.py#L265](https://github.com/home-assistant/home-assistant/blob/177ae3fd322e116f460c3b23d7c0a7a74dc90cd2/homeassistant/components/homematic/__init__.py#L265) and [homeassistant/components/homematic/__init__.py#L281](https://github.com/home-assistant/home-assistant/blob/177ae3fd322e116f460c3b23d7c0a7a74dc90cd2/homeassistant/components/homematic/__init__.py#L281). If I remove the ```socket.gethostbyname``` everything is fine. Is there any reason why an ip-address was forced by using ```socket.gethostbyname```?

My change was just a quick test not meant to be a fix. By changing line [pyhomematic/_hm.py#L515](https://github.com/danielperna84/pyhomematic/blob/8b91f3e84c83f05d289c740d507293a0d6759d8e/pyhomematic/_hm.py#L515) I have reverted the hostname to ip-address conversion from  the Home Assistant component. You see?

If we remove the ```socket.gethostbyname``` from the Home Assistant component the user is free to enter a hostname or an ip-address. But I fear we have to rename the dictionary key ```'ip'``` to ```'host'``` to be consistent.

**Comment by danielperna84** [comment]
> Is there any reason why an ip-address was forced by using `socket.gethostbyname`?

Not that I'm aware of. I didn't even know it's done like that. Probably because way back then when we did this we only accepted IP addresses. 🤷‍♂️ 

Ah, you're right. I didn't read that properly.

Yes, removing the `gethostbyname` is what should be done. I wouldn't rename the keys though. The configuration uses [host](https://www.home-assistant.io/components/homematic/#host) already. Changing just the key for pyhomematic would require an update of the module. That's unnecessary just to fix the problem. So feel free to create a PR for Home Assistant that just removes the `gethostbyname`.
